org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaSemanticTokensProvider.java

```diff
@@ -0,0 +1,162 @@
+			return Collections.emptyList();
+		}
+		
+		IResource resource = ast.getTypeRoot().getResource();
+		if (resource == null) {
+			return Collections.emptyList();
+		}
+		
+		final IDocument theDocument = LSPEclipseUtils.getDocument(resource);
+		if (theDocument == null) {
+			return Collections.emptyList();
+		}
+		
+		SemanticTokensParams params = getSemanticTokensParams(theDocument);
+		if (params == null) {
+			return Collections.emptyList();
+		}
+		
+		LanguageServerDocumentExecutor executor = LanguageServers.forDocument(theDocument)
+				.withFilter(this::hasSemanticTokensFull);
+			
+		try {
+			return executor
+				.computeFirst((w, ls) -> {
+					return ls.getTextDocumentService().semanticTokensFull(params)
+							.thenApply(semanticTokens -> {
+								if (semanticTokens == null) {
+									return Collections.<SemanticToken>emptyList();
+								}
+								SemanticTokensLegend legend = getSemanticTokensLegend(w);
+								if (legend == null) {
+									return Collections.<SemanticToken>emptyList();
+								}
+								return new JavaSemanticTokensProcessor(this::mapToTokenType, p -> {
+									try {
+										return LSPEclipseUtils.toOffset(p, theDocument);
+									} catch (BadLocationException e) {
+										throw new RuntimeException(e);
+									}
+								}).getSemanticTokens(semanticTokens.getData(), legend);
+							});
+				})
+				.thenApply(o -> o.orElse(Collections.emptyList())).get(300, TimeUnit.MILLISECONDS);
+		} catch (InterruptedException | ExecutionException | TimeoutException e) {
+			LanguageServerPlugin.logError("Failed to fetch semantic tokens for '%s' from Language Servers".formatted(resource.getLocation()), e);
+			return Collections.emptyList();
+		}		
+	}
+	
+	private @Nullable SemanticTokensParams getSemanticTokensParams(IDocument document) {
+		URI uri = LSPEclipseUtils.toUri(document);
+		if (uri != null) {
+			final var semanticTokensParams = new SemanticTokensParams();
+			semanticTokensParams.setTextDocument(LSPEclipseUtils.toTextDocumentIdentifier(uri));
+			return semanticTokensParams;
+		}
+		return null;
+	}
+
+	private @Nullable SemanticTokensLegend getSemanticTokensLegend(final LanguageServerWrapper wrapper) {
+		ServerCapabilities serverCapabilities = wrapper.getServerCapabilities();
+		if (serverCapabilities != null) {
+			SemanticTokensWithRegistrationOptions semanticTokensProvider = serverCapabilities
+					.getSemanticTokensProvider();
+			if (semanticTokensProvider != null) {
+				return semanticTokensProvider.getLegend();
+			}
+		}
+		return null;
+	}
+
+	private boolean hasSemanticTokensFull(final ServerCapabilities serverCapabilities) {
+		return serverCapabilities.getSemanticTokensProvider() != null
+				&& LSPEclipseUtils.hasCapability(serverCapabilities.getSemanticTokensProvider().getFull());
+	}
+
+	private ISemanticTokensProvider.TokenType mapToTokenType(String tokeTypeStr) {
+		switch (tokeTypeStr) {
```

Comment:
Could we use the syntax
return switch (tokenTypeStr) {
  case "method" -> ISemanticTokensProvider.TokenType.METHOD
  ...